### PR TITLE
Initial integration with Decidim proposals

### PR DESCRIPTION
**Pull request by jinjagit**
~~Continues work on [#2](https://github.com/liquidvotingio/decidim-instance/issues/2)~~

**Update** (Oliver)

Closes https://github.com/liquidvotingio/decidim-module-liquidvoting/issues/27

### CODE REVIEW

**Comment by jinjagit** [comment]
Weirdness detected in original `app/views/decidim/proposals/proposals/_vote_button.html.erb`:  
This conditional...
```
<% if component_settings.participatory_texts_enabled? && from_proposals_list %>
    <%= render partial: "decidim/proposals/proposals/participatory_texts/proposal_vote_button", locals: { proposal: proposal, from_proposals_list: true } %>
  <% else %>
```
If `true`, seems to destroy the layout of the endorse button and associated elements:  
weirdness
I haven't yet understood why this happens,, nor what to do about it.

**Comment by oliverbarnes** [comment]
No need to add WIP to the title when the PR is set to Draft. I've already edited it accordingly

**Comment by jinjagit** [comment]
Useful Decidim graph of proposal life-cycle:  
proposal

**Comment by jinjagit** [comment]
Note on Participatory Texts:  
To view a process that contains participatory texts (and probably any type of process), the admin needs to set a start/end date for the process (and this probably needs to include the current date).

**Comment by jinjagit** [comment]
Args passed to `decidim-core/app/helpers/decidim/action_authorization_helper.rb/action_authorized_button_to(action, *arguments, &block)`...  
  
When creating a vote:  
```
action: vote

*arguments: [
  "/processes/rerum-eaque/f/12/proposals/14/proposal_vote?from_proposals_list=false",
    {:resource=>#<Decidim::Proposals::Proposal id: 14, title: "Nisi quisquam.",
      body: "Ullam cumque ad. Voluptas incidunt mollitia. In om...",
      decidim_component_id: 12,
      decidim_scope_id: nil,
      created_at: "2020-10-26 16:30:37", updated_at: "2020-10-26 16:30:37",
      proposal_votes_count: 1,
      state: "accepted",
      answered_at: "2020-10-26 16:30:37",
      answer: {"ca"=>"Nihil consequatur corrupti molestiae voluptatem sunt reiciendis ut excepturi sed.",
                "en"=>"Alias quasi eos est corporis et eum quae non quisquam.",
                "es"=>"Aspernatur est ut fugit vel vel quod consequatur natus consectetur."},
      reference: "DVM-PROP-2020-10-14",
      address: nil,
      latitude: nil,
      longitude: nil,
      proposal_endorsements_count: 0,
      published_at: "2020-10-26 16:30:37",
      proposal_notes_count: 1,
      coauthorships_count: 5,
      position: 7,
      participatory_text_level: nil,
      created_in_meeting: false,
      endorsements_count: 8,
      cost: nil,
      cost_report: nil,
      execution_period: nil,
      state_published_at: "2020-10-26 16:30:37">,
      :remote=>true,
      :data=>{:disable=>true,
              :"redirect-url"=>"/processes/rerum-eaque/f/12/proposals/14"},
      :class=>"button expanded"}
]

&block: #<Proc:0x00007f2d88f07f50 /home/j/Documents/liquid/decidim-liquidvoting/app/views/decidim/proposals/proposals/_vote_button.html.erb:66>
&block params: []
```
  
When deleting a vote:  
```
action: vote

*arguments: [
  "/processes/rerum-eaque/f/12/proposals/14/proposal_vote?from_proposals_list=false",
  {:resource=>#<Decidim::Proposals::Proposal id: 14,
    title: "Nisi quisquam.",
    body: "Ullam cumque ad. Voluptas incidunt mollitia.
    In om...", decidim_component_id: 12,
    decidim_scope_id: nil,
    created_at: "2020-10-26 16:30:37",
    updated_at: "2020-10-26 16:30:37",
    proposal_votes_count: 2,
    state: "accepted",
    answered_at: "2020-10-26 16:30:37",
    answer: {"ca"=>"Nihil consequatur corrupti molestiae voluptatem sunt reiciendis ut excepturi sed.",
              "en"=>"Alias quasi eos est corporis et eum quae non quisquam.",
              "es"=>"Aspernatur est ut fugit vel vel quod consequatur natus consectetur."},
    reference: "DVM-PROP-2020-10-14",
    address: nil,
    latitude: nil,
    longitude: nil,
    proposal_endorsements_count: 0,
    published_at: "2020-10-26 16:30:37",
    proposal_notes_count: 1,
    coauthorships_count: 5,
    position: 7,
    participatory_text_level: nil,
    created_in_meeting: false,
    endorsements_count: 8,
    cost: nil, cost_report: nil,
    execution_period: nil,
    state_published_at: "2020-10-26 16:30:37">,
    :method=>:delete,
    :remote=>true,
    :data=>{:disable=>true,
            :original=>"Already supported",
            :replace=>"Withdraw support",
            :"redirect-url"=>"/processes/rerum-eaque/f/12/proposals/14"},
    :class=>"button expanded active light",
    :id=>"vote_button-14"}
]

&block: #<Proc:0x0000556ce7999450 /home/j/Documents/liquid/decidim-liquidvoting/app/views/decidim/proposals/proposals/_vote_button.html.erb:51>
&block params: []
```

**Comment by oliverbarnes** [comment]
> Args passed to `decidim-core/app/helpers/decidim/action_authorization_helper.rb/action_authorized_button_to(action, *arguments, &block)`...

What's the story with these args? I'm a little lost here

**Comment by jinjagit** [comment]
These are the args that are passed to a function in decidim-core, when the original 'Support' button is clicked.

**Comment by jinjagit** [comment]
Note on 'Support' button disappearing when clicked:   
  
The 'Support' button disappears when clicked, but when the page is manually refreshed it has updated to the 'Already supported' state. This also happens if the button is then subsequently clicked again (returning to it's original state on manual page refresh). I think this has always been the case for the `decidim-liquidvoting` repo.  
Removing all of our code from the _vote_button partial, so that it is exactly the same as the original in decidim/proposal, does not correct this behavior.  
The intended functionality is for the button to update immediately it is clicked, without manual page refresh (confirmed by testing on [meta-decidim website](https://meta.decidim.org)).

**Comment by oliverbarnes** [comment]
Yes, we'll need an ajax request to get there. I think this was discussed at some point - we could start with a page refresh and then implement the async request with js after that, following whatever is the latest decidim is using for these. Btw, React [was dropped recently](https://github.com/decidim/decidim/pull/6498), which makes things a lot easier js-wise.

**Comment by jinjagit** [comment]
Ah, I see. I didn't realize _that_ was the reason we were considering using ajax.

File: app/overrides/commands/decidim/proposals/vote_proposal.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        #   @proposal.with_lock do
+        #     vote.save!
+        #     update_temporary_votes
+        #   end
+        # end
+
+        Decidim::Liquidvoting::Client.create_vote(
+          # NOTE: This is a clunky way of creating the url.
+          # Probably can find a better way to pass the url here.
+          # Also, what is the '.../f/...' part of the url?
```
Comment (oliverbarnes) [inline app/overrides/commands/decidim/proposals/vote_proposal.rb:39]: Looking at the `Proposal` model with hopes of finding a reference to its url, I found this:

```ruby
# Builds the url to the resource. Useful when linking to a resource from
# another engine.
#
# options - An optional hash of options to pass to the Rails router
#
# Returns a String.
def url(options = {})
  member_route("url", options.merge(host: root_resource.organization.host))
end
```

The `ResourceLocatorPresenter#url` method def:

```ruby

```

File: app/overrides/controllers/decidim/proposals/proposal_votes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      def create
+        enforce_permission_to :vote, :proposal, proposal: proposal
+        @from_proposals_list = params[:from_proposals_list] == "true"
+
+        VoteProposal.call(proposal, current_user) do
+          on(:ok) do
+            proposal.reload
+
+            # TODO: Replace with search that uses LV proposals to select which proposals to include.
+            proposals = ProposalVote.where(
```
Comment (oliverbarnes) [inline app/overrides/controllers/decidim/proposals/proposal_votes_controller.rb:23]: LV proposals?

Comment (jinjagit) [reply app/overrides/controllers/decidim/proposals/proposal_votes_controller.rb:23]: It means making use of the `proposal_url` stored with votes (or possibly delegations) on the LV db.

Comment (oliverbarnes) [reply app/overrides/controllers/decidim/proposals/proposal_votes_controller.rb:23]: ok, I see

Comment (jinjagit) [reply app/overrides/controllers/decidim/proposals/proposal_votes_controller.rb:23]: The url contains the `component.participatory_space` ID & the `proposal` id, as part of the url string, and thus can be used to filter proposals on the decidim side.

File: app/overrides/commands/decidim/proposals/vote_proposal.rb (hunk 1 above)
Comment (davefrey) [reply app/overrides/commands/decidim/proposals/vote_proposal.rb:39]: There's tons of this in the decidim repo, including precedent for using from models, here's an example usage:
```Decidim::ResourceLocatorPresenter.new(proposal).url```

The url is also available in the LV state object `@lv_state.proposal_url`

Comment (davefrey) [reply app/overrides/commands/decidim/proposals/vote_proposal.rb:39]: There are also two calls to `Decidim::Liquidvoting::Client.create_vote`, I'm unsure yet if both are needed. 

One is this one, in `ProposalVote.call`, the other in `VoteController.create`.

Comment (oliverbarnes) [reply app/overrides/commands/decidim/proposals/vote_proposal.rb:39]: > There's tons of this in the decidim repo, including precedent for using from models, here's an example usage:
> `Decidim::ResourceLocatorPresenter.new(proposal).url`
> 
> The url is also available in the LV state object `@lv_state.proposal_url`

Yeah, not sure we need to mirror it in the LV state. I'd say keep only LV-only or LV-modified data there. Our integration doesn't affect proposal_url

**Comment by oliverbarnes** [comment]
Renamed the PR to match the tracking issue: https://github.com/liquidvotingio/decidim-module-liquidvoting/issues/27

I'd like to merge this as soon as it's minimally solid to avoid another never-ending PR, and after that we can increment piecemeal (like with the LV state stuff, which would have been a nice discrete PR).

File: app/overrides/commands/decidim/proposals/vote_proposal.rb (hunk 1 above)
Comment (oliverbarnes) [reply app/overrides/commands/decidim/proposals/vote_proposal.rb:39]: > There are also two calls to `Decidim::Liquidvoting::Client.create_vote`, I'm unsure yet if both are needed.
> 
> One is this one, in `ProposalVote.call`, the other in `VoteController.create`.

It'd be create to have a single place where te API is called for this.

Comment (davefrey) [reply app/overrides/commands/decidim/proposals/vote_proposal.rb:39]: Agreed, if both entries are needed we can perhaps pass both through the command object...I'll get it sorted.

Comment (davefrey) [reply app/overrides/commands/decidim/proposals/vote_proposal.rb:39]: Also agree for the things we don't need in the state object (email and url are needed to *retrieve* the LV info, but don't need to be carried with the state object). I've just removed those.

**Comment by davefrey** [comment]
👋  This PR is so too long, I propose we review, break out new issues for work we still want to do, and get it merged. The work is in a relatively consistent state, and does offer a working, coherent delegation/voting UI, overriding Decidim's proposal voting and abandoning their `ProposalVote` model.

To do this, I suggest we:
1. Unlink issue #18 (don't Close it here) -- the vote part is done, delegations are working as page requests, ajax for this has been problematic - we can manage that separately on #18
2. Make the other scope reductions I've suggested on issue #27

**Review by oliverbarnes — APPROVED** [review]
Approving so we can merge this big-bang PR, which is functional at least, and move on into smaller discrete tasks

**Pull request merged**